### ISSUE TITLE
Add Flush method to the SpriteBatch

### ISSUE THREAD

**Issue by rds1983** [MonoGame/MonoGame#5241]
Right now, if one wants to send the SpriteBatch's draw data to the graphics device and continue to draw (i.e. if he changed the ScissorRectangle), then he needs to call the End method followed by Begin. That could be a performance hit(and actually it is according to my testing) as every End/Begin reapplies the same graphics states over and over again.
An explicit Flush method would solve that issue.

**Comment by nkast** [comment]
What ShortMode you used, Deffered, Immediate?
Imediate is for senarios like that. It flash on every Draw.

 how you tested it and what diff did you see?

**Comment by rds1983** [comment]
I develop a GUI Library:
![image](https://cloud.githubusercontent.com/assets/1057289/19677716/c346fca2-9ac4-11e6-940a-d278e6703177.png)

Initially I used Deffered SortMode and called End/Begin after changing the ScissorRectangle . However that approach didnt work smooth in terms of performance (it went below 60 noticeably and there were even freezes). Therefore I switched to Immediate. It works well. However as you can see on screenshot, it does a draw call for every character. I though it would be good if sprite drawing could be batched for every widget.

**Comment by nkast** [comment]
> Therefore I switched to Immediate. It works well. However as you can see on screenshot, it does a draw call for every character.

Actually DrawString() doesn't flash on every character (AutoFlash is false). But it's true that the Immediate mode flash the buffer on every Draw/DrawString(). It's surprising that Immediate is faster than End()/Begin().  A flash apply the chances on the DeviceContext, while End() just set's 'IsXDirty' flags.
I prefer to first investigate for bottlenecks before changing the API. My bet is on Begin() and the null checks.

Meanwhile, can you check #5226 and see it it's an improvement in your case?
http://teamcity.monogame.net/repository/download/MonoGame_PackagingWindows/26135:id/MonoGameSetup.exe

**Comment by rds1983** [comment]
I've attached a small sample to demonstrate an issue.
![image](https://cloud.githubusercontent.com/assets/1057289/19683964/edc5ffcc-9ade-11e6-9b3e-6cfae36542ba.png)
It simply draws 500 "Hello, World!"s using 3 different modes: BeginEnd, FlushReflection and Immediate, which can be switched beetween by pressing the Space Key.
First mode calls End()/Begin() after every DrawString and it results into obvious freezes(as seen on screenshot FPS fell to ~20). Btw Begin() is not parameterless, but it has following look:
`_spriteBatch.Begin( SpriteSortMode.Deferred, BlendState.AlphaBlend, SamplerState.PointClamp, null, new RasterizerState
			{
				ScissorTestEnable = true
			});`

Second mode emulates proposed Flush(it invokes _batcher.DrawBatch(_sortMode, _effect) using the reflection) after every DrawString.
Finally, third mode use Immediate SpriteSortMode.
Last two modes(second and third) doesnt have any issues with the performance. 

[SpriteBatchTest.zip](https://github.com/MonoGame/MonoGame/files/550422/SpriteBatchTest.zip)

**Comment by nkast** [comment]
> new RasterizerState { ScissorTestEnable = true }

Initialize a RasterizerState once in the LoadContent() (or in Initialize()) and use that for the parameter.
Try to avoid creating State objects on every frame.

**Comment by james0x0A** [comment]
I gave this a test, and indeed  creating the RasterizerState  several times per frame causes serious stuttering.   

 Using a single allocation of the RasterizerState  in LoadContent results in smooth frame rates.  Still, it is notable that on my system using DesktopGL the reflection based flush method gave a 15% performance improvement over calling End/Begin  many times for a small workload. And Immediate was faster  by at least another 20%.  I assume the difference will be eroded as the workload of each batch grows, but its there.

**Comment by nkast** [comment]
I run a test before updating my branch. I moved RasterizerState initialization to LoadContent and disable FixedTimeStep & VSync.
```
            this.IsFixedTimeStep = false;
            this._graphics.SynchronizeWithVerticalRetrace = false;
```
Initially I got: BeginEnd 148fps, Flush 182fps, Immediate 213fps.
With #5226, Immediate went up to 228fps, a 7% improvement! /play trombone
There were no noticeable improvements with BeginEnd and Flush.

**Comment by james0x0A** [comment]
I moved all unnecessary per frame object creation from the draw call and I saw slight bump using the latest with your improvements. Probably around that 7% you mentioned.

**Comment by nkast** [comment]
Yeap. I just moved creation of the reflection fields into LoadContent and reflectionFlush went up to 206fps-213fps.

**Comment by rds1983** [comment]
Thanks for the advices. I've also moved RasterizerState and reflection variables into class fields and got same results(it runs smooth, Immediate is faster than Flush which is faster than BeginEnd).
However I've also made it output draw calls and found out that Immediate does almost same amount as Flush. Which means Immediate DrawString batches it.
So I've turned off that batching by making it draw each symbol separately.
Finally it has expected results (Flush is slightly faster than BeginEnd which is significally faster than Immediately).
Source:
https://gist.github.com/rds1983/bc20059fbea4f56cd6ce7ba5919f250f

In my GUI library I dont use SpriteFont as you can't build it dynamically(and I dont use Content Pipeline). So I use my own BitmapFont and the text rendering code. Therefore I still support idea of adding Flush method to the SpriteBatch.

**Comment by nkast** [comment]
>  I use my own BitmapFont and the text rendering code and so each symbol is drawn separately
Oh, i see.

I profiled EndBegin and ReflectionFlush and found that the time is lost in Setup method. The hotspots are the projection calculation and applying the SpriteEffect. 

I am thinking that a Flush() can't avoid calling Setup() if we don't want to break the expected behavior. With ImmediateMode Flush() doesn't have any effect. With the other sorting methods you expect things to be drawn with the states you use to initialize Begin() and not be overridden by nested SpriteBatches or other Drawing operations. Another senario similar to  ScissorRectangle is setting the Viewport. In that case it has to update the projection matrix on every Flush(), right?

**Comment by nkast** [comment]
Here is the difference between ReflectionFlush & EndBegin. I disable FixedTimeStep and Vsync for tihs test. 4.36% of the time is wasted in Setup(), which is the difference between DrawIntro (25.21%-18.93%=6.28%).

![endbegin](https://cloud.githubusercontent.com/assets/3018589/19712401/15777cd2-9b46-11e6-94a3-424aaee096b0.png)

Some of this overhead can be avoided. For example multiplying the projection with Identity when parameter transform is null, which is almost always the case.
Calculating the projection and applying the skinnedEffect can be skipped when user pass it's own effect in Begin(). FlushVertices() uses the user defined effect in that case. 

With those fixes you can define your own effect/projection and reuse it with begin(). The fact is that SpriteBatch wasn't design around the scenario of calling Begin() .End() frequently. Changing ScissorRectangle requires a stage change and that's always expensive.

**Comment by rds1983** [comment]
I dont think there's point of calling Setup every Flush as it wont be any perfomance boost over calling End/Begin. I think Setup should be called once per Begin/End cycle. It could be implemented by adding _setupCalled boolean field to the SpriteBatch. And the Flush code would look like:
`public void Flush()
{
  if (!_setupCalled) {
    Setup();
    _setupCalled = true;
  }

  ... Flush code
}`
So if a user calls Flush multiple times during Begin/End, the Setup is called only once.
Of course, that will bring issues if a user uses nested SpriteBatches or changes Viewport.
But firstly, there are already such issues if Immediate SpriteSortMode is used.
Secondly I think the Setup method could become public as well. So if a user has multiple SpriteBatches he can fully control when their states are applied.

**Comment by nkast** [comment]
#5244 will give you a 15%~20% on End()/Begin(). Other modes are unaffected. Immediate is still faster on the TestSample, but I don't know in your case where you have to batch many charachers. 

You have to pass a custom effect like the one bellow:
```
            var vp = GraphicsDevice.Viewport;
            Matrix projection;            
            Matrix.CreateOrthographicOffCenter(0, vp.Width, vp.Height, 0, 0, -1, out projection);
            projection.M41 += -0.5f * projection.M11;
            projection.M42 += -0.5f * projection.M22;
            _effect = new BasicEffect(GraphicsDevice);
            _effect.TextureEnabled = true;
            _effect.VertexColorEnabled = true;
            _effect.Projection =projection;
            _effect.View = Matrix.Identity;
            _effect.World = Matrix.Identity;
```

By the way, you can subclass ContentManager and override one method to fool it and give it a Memory stream. That way you can pass a spitefont through the Load<>() method.

**Comment by nkast** [comment]
>Of course, that will bring issues if a user uses nested SpriteBatches or changes Viewport.
>But firstly, there are already such issues if Immediate SpriteSortMode is used.

Exactly. ex.
sb.Begin(Deffered, customBlend1);
device.BlendState = customBlend2;
sb.Draw(something);
sb.Flash(); // draw something with customBlend1
device.BlendState = customBlend3;
sb.Draw(somethingElse);
sb.End(); // draw something with customBlend3

It doesn't make any scene. The first flash() acts as DefferedMode, after that it behaves as Immediate.

**Comment by rds1983** [comment]
Well, new set of Modes is great idea. If I understand you correctly it would work like following:
sb.Begin(SpriteSortMode.Flush, ...); // Setup is called here
sb.Draw(something);
sb.Draw(something);
sb.Draw(something);
sb.Flush(); // Batch is sent to the graphics device (probably should throw an exception it called within one of older Modes)
sb.Draw(something);
sb.Draw(something);
sb.Draw(something);
sb.Flush(); // Batch is sent to the graphics device
sb.Draw(something);
sb.Draw(something);
sb.End(); // Batch is sent to the graphics device and the SpriteBatch is closed.

That would definitely solve my case.

**Comment by nkast** [comment]
I am worried that a Flush() method will cause more confusion to end users because it tries to solve a performance issues on a very rare scenario, the same functionality can be simulated with End()/Begin(). 
It wont do anything on the current SortModes. The slow down with End()/Begin() is re-applying the SpriteEffect, which is required at the moment. Another approach would be a begin overload with a boolean 'dontApplyEffect' parameter, preferably at the same position where effect currently is. that change will serve no purpose if we find a way to skip the effect initialization. We can probably do that for stock Effects like `BasicEffect`. 

What fps you get on your GUI library with EndBegin, ReflectionFlush and Immediate? The sample you gave does a poor job to present the problem because it uses DrawString and Immediate looks like the best option. A test that draws many textures will be better to see if there a real perf issue.

Anyways, when max performance is required a custom solution work best, what's stopping you from grabbing SpriteBatch.cs and fine-tune it to your needs? It's 'Microsoft Public License (Ms-PL)'.

**Comment by rds1983** [comment]
Well, any approach will work. Also SpriteBatch could be split into SpriteBatchBase and actual SpriteBatch(which inherits from SpriteBatchBase). SpriteBatchBase would provide basic functionality for batching sprites and most importantly it'll have protected methods Setup and Flush. That approach wont change API of the original SpriteBatch, but will allow end users to inherit from SpriteBatchBase and write their own sprite batches  designed for their needs.

I've posted updated sample a bit earlier that does separate draw call for each symbol and it clearly demonstrates that ReflectionFlush gives best performance:
https://gist.github.com/rds1983/bc20059fbea4f56cd6ce7ba5919f250f

Of course, if we're talking about my case - there are thousands of workaround. From using ReflectionFlush to writing my own SpriteBatch. But I thought that others may benefit flushable sprite batches as well.

**Comment by nkast** [comment]
> Also SpriteBatch could be split into SpriteBatchBase and actual 
> SpriteBatchBase would provide basic functionality for batching sprites and most importantly it'll have protected methods Setup and Flush

Good idea, seperate the batching from drawing and provide an abstract/base class. I think we have to do some refactoring first to come up with a solid interface. I would also add to that, the ability to pre-batch a layout, string, etc and feed that to the SpriteBatch.

I run you test and indeed, immediate is way too slow, but EndBegin is not too far from ReflectiveFlush.
I tried a few thing to improve Begin/End perf but there's no way around it. SpriteBatch is designed to be safe.

Even with a Flush method or custom SpriteBatch you are still running slow code because of changing the Scissor too often. How your tried other approaches? For example you could turn drawing off and draw rectangles to the depth buffer for each widget with an increasing layer value (or decreasing, you have to check which one works). Then draw your widget in the same order and with the same depth as before. This way you avoid reseting the states and also get lower pixelcount, as most draws will fail the depth check.

**Comment by nkast** [comment]
There is a CompareFunction.Equal you can set DepthStencilState.DepthBufferFunction to, and draw each widget with a depth as it's unique id.
https://msdn.microsoft.com/en-us/library/microsoft.xna.framework.graphics.comparefunction.aspx

**Comment by nkast** [comment]
@rds1983 FYI, #5456. 
I hope it somewhat improves your scenario.

**Comment by rds1983** [comment]
Awesome fix, thanks!

**Comment by rds1983** [comment]
Meanwhile I've taken a look at Xenko's SpriteBatch:
https://github.com/SiliconStudio/xenko/blob/master/sources/engine/SiliconStudio.Xenko.Graphics/SpriteBatch.cs

And found out that it inherits from BatchBase (though Flush method is private for some reason):
https://github.com/SiliconStudio/xenko/blob/master/sources/engine/SiliconStudio.Xenko.Graphics/BatchBase.cs

So separating batching from drawing seems to be a working solution. I can try to bring it to the MonoGame and prepare the pull request if that idea gets approved.

**Comment by rds1983** [comment]
It's so helpful to look at others sources.
As I've looked at how Xenko developers solved problem with widgets clipping:
https://github.com/SiliconStudio/xenko/blob/master/sources/engine/SiliconStudio.Xenko.UI/Rendering/UI/UIRenderFeature.cs
(see method RecursiveDrawWithClipping)

And found out that they go with @nkast 's proposal; using depth buffer instead of scissors. That makes a lot of sense as that approach allows to clip transformed(i.e. rotated) widgets. 
However most interested thing is that they dont clip all widgets, but ones with ClipToBounds set to true.
Further analysis revealed that there's only one such widget - ScrollViewer. And again that makes a lot of sense.

So probably I'll simply copy their approach and clip only ScrollPane(my version of ScrollViewer). That'll solve my case as calling Begin/End once per ScrollPane isn't a problem performance-wise.

Though my proposition in the previous post is still actual. I am ready to split SpriteBatch into SpriteBatchBase/SpriteBatch and make a pull request.
